---
title: "Compare DE methods for 10X genomics - edgeR and us"
author: "Fanny Perraudeau"
date: "`r Sys.Date()`"
output: 
html_document: 
fig_height: 7
fig_width: 7
toc: yes
code_folding: hide
toc_float: yes
---

```{r options, echo=FALSE, results="hide",mesasge=FALSE, error=FALSE, include=FALSE, autodep=TRUE}
knitr::opts_chunk$set(fig.align="center", cache=TRUE, error=FALSE, message=FALSE, warning=TRUE)
library(zinbwave)
library(edgeR)
library(zingeR)
library(RColorBrewer)
library(ggplot2)
```


After running Seurat preprocessing (filtering, normalization, dimensionality reduction, clustering), we want to find DE genes between clusters using Seurat DE tool, edgeR, edgeR with our weights. Then, we'll perform gene enrichment analysis.

```{r loadobject}
load("core.rda")
core
```

### Finding differentially expressed genes (cluster biomarkers)
```{r fit}
computeZinbwaveWeights <- function(zinb, counts){
  mu <- getMu(zinb)
  pi <- getPi(zinb)
  theta <- getTheta(zinb)
  theta_mat <- matrix(rep(theta, each = ncol(counts)), ncol = nrow(counts))
  nb_part <- dnbinom(t(counts), size = theta_mat, mu = mu)
  zinb_part <- pi * ( t(counts) == 0 ) + (1 - pi) *  nb_part
  zinbwg <- ( (1 - pi) * nb_part ) / zinb_part 
  t(zinbwg)
}
```

```{r,eval=FALSE}
library(doParallel)
library(BiocParallel)
NCORES = 2
registerDoParallel(NCORES)
register(DoparParam())
print(system.time(zinb <- zinbFit(core, X = '~ 0 + seuratFactor',
                                  epsilon = 1e8)))
save(zinb, file = 'zinb.rda')
```

```{r weightszinb}
load('zinb.rda')
counts = assay(core)
weights_zinbwave = computeZinbwaveWeights(zinb, counts)
hist(weights_zinbwave, main = 'zinbwave', xlab = 'Weights')
```

# Design matrix

```{r zinbde}
clus = colData(core)$seuratVector
nclus = length(unique(clus))
design <- model.matrix(~ 0 + colData(core)$seuratFactor)
contrast = matrix(-1/7, nrow = nclus, ncol = nclus)
diag(contrast) = 1
colnames(contrast) = paste0('c', 1:length(unique(clus)))
```

# Run edgeR
```{r fitedger}
fit_edgeR <- function(counts, design, contrast, filter = NULL){
  library(edgeR)
  d = DGEList(counts)
  d = suppressWarnings(calcNormFactors(d))
  d = estimateDisp(d, design)
  fit = glmFit(d,design)
  
  de = lapply(1:ncol(contrast), function(i){
    glm = glmLRT(fit, contrast = contrast[, i])
    tab = glm$table
    tab$padj = p.adjust(tab$PValue, "BH")
    tab$contrast = colnames(contrast)[i]
    tab$gene = rownames(tab)
    tab
  })
  de <- do.call(rbind, de)
  de <- data.frame(de, stringsAsFactors = FALSE)
  de = de[, c('gene', 'contrast', 'PValue', 'padj', 'logFC')]
  colnames(de) = c('gene', 'contrast', 'pval', 'padj', 'logfc')
  de
}
```

```{r edgeR,eval=FALSE}
edgeR <- fit_edgeR(counts, design, contrast)
edgeR$method <- 'edgeR'
```

# Run edgeR with zinbwave weights
```{r fitedgeRzi}
fit_edgeR_zi <- function(counts, design, contrast, weights,
                         filter = NULL){
  library(edgeR)
  library(zingeR)
  d = DGEList(counts)
  d = suppressWarnings(calcNormFactors(d))
  d$weights <- weights 
  d = estimateDisp(d, design)
  fit = glmFit(d,design)
  
  de = lapply(1:ncol(contrast), function(i){
    glm = glmWeightedF(fit, contrast = contrast[, i], 
                       filter = filter)
    tab = glm$table
    tab$contrast = colnames(contrast)[i]
    tab$gene = rownames(tab)
    tab
  })
  de <- do.call(rbind, de)
  de <- data.frame(de, stringsAsFactors = FALSE)
  de = de[, c('gene', 'contrast', 'PValue', 'padjFilter', 'logFC')]
  colnames(de) = c('gene', 'contrast', 'pval', 'padj', 'logfc')
  de
}
```

```{r zinbwaveedgeR,eval=FALSE}
# check if filter ok with baseMean, not sure
nf <- edgeR::calcNormFactors(counts)
baseMean = unname(rowMeans(sweep(counts,2,nf,FUN="*")))
zinbwave_edgeR <- fit_edgeR_zi(counts, design, contrast, 
                            weights = weights_zinbwave,
                            filter = baseMean)
zinbwave_edgeR$method <- 'zinbwave_edgeR'
```

```{r,eval=FALSE}
de = rbind(edgeR, zinbwave_edgeR)
save(de, file = 'de.rda')
```

```{r}
load('de.rda')
head(de,2)
```

# Seurat
```{r}
seurat = read.csv('seurat_de_res.csv')
seurat$padj = NA
seurat$method = 'seurat'
seurat = seurat[, c('gene', 'cluster', 'p_val', 'padj', 'avg_diff', 'method')]
colnames(seurat) = colnames(de)
seurat$contrast = paste0('c', seurat$contrast + 1)
seurat = seurat[seurat$gene %in% unique(de$gene), ]
```

```{r}
de = rbind(de, seurat)
```

# Compare methods
## Histogram of pvalues

We are expecting uniformity of the pvalue with a pick close to zero corresponding to DE genes.

```{r hist}
ggplot(de, aes(pval)) + geom_histogram() + 
  facet_grid(contrast ~ method)
```

## Volcano plots
```{r}
for (y in unique(de$contrast)){
  par(mfrow=c(1,3))
  for (x in unique(de$method)){
    temp = de[de$method == x & de$contrast == y, ]
    plot(temp[, 'logfc'], -log10(temp[, 'pval']),
         pch = 20, col = 'gray', cex = .5, xlim = c(-10, 10), ylim = c(0, 50),
         ylab = '-log10(pvalue)', xlab = 'logFC', main = paste(y,'\n',x))
  }
}
par(mfrow=c(1,1))
```

## Concordance between DE genes
### Venn diagram

```{r}
thr = 0.05
```

DE genes are genes with an adjusted pvalue lower than `r thr`.

```{r venn,eval=FALSE}
library(dplyr)
ve = de %>% group_by(method, gene) %>%
  summarize(pval = min(pval, na.rm=TRUE)) %>%
  ungroup() %>% as.data.frame()
ve$de = ve$pval < thr
ve = reshape2::acast(ve[, c(1,2,4)], gene ~ method, sum)

aa <- vennCounts(ve)
vennDiagram(aa, main = 'DE gene, adj pvalue < 0.05')
```


## Most DE genes per contrast and method
```{r}
ranks = lapply(unique(de$contrast), function(y){
  lapply(unique(de$method), function(x){
    temp = de[de$method == x & de$contrast == y, ]
    temp$rank = rank(temp$pval)
    temp
  })
})
de = do.call(rbind, do.call(rbind, ranks))

ranks = lapply(unique(de$contrast), function(y){
  permethod = lapply(unique(de$method), function(x){
    temp = de[de$method == x & de$contrast == y, ]
    temp = temp[temp$rank < 10, ]
    temp[order(temp$rank), ]
  })
  names(permethod) = unique(de$method)
  permethod
})
names(ranks) = unique(de$contrast)
ranks
```

## Heatmaps 

```{r tcnorm}
norm10x <- function (ei){
  sums = colSums(ei)
  t(t(ei)*median(sums)/sums)
}
tc = norm10x(counts) 
tc = log1p(tc)
```

```{r heatmaptc}
library(clusterExperiment)
tt = lapply(unique(de$method), function(x){
  temp = de[de$method == x, ]
  deList = lapply(unique(temp$contrast), function(y){
    temp[order(temp[temp$contrast == y, 'pval'])[1:10], 'gene']
  })
  clusters = as.numeric(colData(core)$seuratNames)
  plotHeatmap(tc[rownames(tc) %in% unique(unlist(deList)), ],
              main = x, breaks = .99,
              sampleData = data.frame(clusters = clusters))
})
```

## Visualize clustering with DE genes

DE genes are genes with an adjusted pvalue lower than `r thr`. We subset the data to keep only DE genes and visualize the clusters in 2D.

```{r pcaDE}
library(rARPACK) 
library(Rtsne)
fastpca <- function(expr, scale=FALSE, k = 50) {
  svd_raw <- svds(scale(t(expr), center=TRUE, scale=scale),
                  k=k, nu=k, nv=0)
  pc_raw <- svd_raw$u %*% diag(svd_raw$d[1:k])
  return(pc_raw)
}

lapply(unique(de$method), function(x){
  temp = de[de$method == x, ]
  degenes = temp[order(temp$pval), 'gene'][1:1000]
  pca = fastpca(tc[rownames(tc) %in% degenes, ])
  pca_tsne <- Rtsne(pca, pca = FALSE)
  plot(pca_tsne$Y, pch=19, col=as.integer(clus))
})
```


## GSEA pre ranked

```{r}
library(xCell) # for db
library(fgsea)
library(GSEABase)

## extract genesets from xcell
genesets <- unlist(geneIds(xCell.data$signatures))
celltypes <- sapply(strsplit(names(genesets), "%"), function(x) x[1])
names(genesets) <- NULL
gs <- tapply(genesets, celltypes, c)


lapply(unique(de$method), function(x){
  temp = de[de$method == x, ]
  lapply(unique(temp$contrast), function(y){
    temp = temp[temp$contrast == y, ]
    degenes = temp$logfc
    names(degenes) = temp$gene
    gsea = fgsea(gs, degenes, nperm = 1000, minSize = 5)
    gsea = gsea[order(gsea$pval), ]
    head(gsea)
  })
})
```

