---
title: "Compare DE methods for 10X genomics - seurat"
author: "Fanny Perraudeau"
date: "`r Sys.Date()`"
output: 
html_document: 
fig_height: 7
fig_width: 7
toc: yes
code_folding: hide
toc_float: yes
---

```{r options, echo=FALSE, results="hide",mesasge=FALSE, error=FALSE, include=FALSE, autodep=TRUE}
knitr::opts_chunk$set(fig.align="center", cache=TRUE, error=FALSE, message=FALSE, warning=TRUE)
library(Seurat)
```

After running Seurat preprocessing (filtering, normalization, dimensionality reduction, clustering), we want to find DE genes between clusters using Seurat DE tool, edgeR, edgeR with our weights. Then, we'll perform gene enrichment analysis.

# Seurat

File pbmc_tutorial.Robj was generated running the pbmc tutorial from Seurat http://satijalab.org/seurat/get_started.html

```{r}
load('../seurat_tuto/pbmc_tutorial.Robj')
```

Seurat has four tests for differential expression which can be set with the test.use parameter: ROC test ("roc"), t-test ("t"), LRT test based on zero-inflated data ("bimod", likelihood-ratio test for single cell gene expression, McDavid et al., Bioinformatics, 2013, default), LRT test based on tobit-censoring models ("tobit").

```{r markers1, fig.height=8, fig.width=15, warning=FALSE}
pbmc.markers <- FindAllMarkers(object = pbmc, min.pct = 0,
                               min.diff.pct = -Inf,
                               max.cells.per.ident = Inf,
                               print.bar = FALSE,
                               genes.use = rownames(pbmc@raw.data),
                               thresh.use = 0, test.use = 'bimod',
                               return.thresh = Inf, min.cells = 0,
                               latent.vars = "nUMI")
length(unique(pbmc.markers$gene))
nrow(pbmc@raw.data)
```

```{r}
library(dplyr)
head(pbmc.markers, 2)
write.csv(pbmc.markers, file = 'seurat_de_res.csv')
topgenes = pbmc.markers %>% group_by(cluster) %>% top_n(2, avg_diff)
topgenes
```


