# Data analysis for the ZINB-WaVE / zingeR paper

This repository is designed to allow interested people to reproduce the results and figures of our paper called 'Unlocking RNA-seq tools for zero inflation and single-cell applications using observation weights'.   

For any questions or issues with the code on this repository, please use the "Issues" tab.

## Dependencies

To be able to run the code in this repo, it is required to have `R` (>=3.4) and the following packages.

### R packages

- [zingeR](https://github.com/statOmics/zingeR)  
- Seurat  
- ggplot2  
- RColorBrewer  
- devtools  
- scales  
- gridBase  
- grid  
- Matrix  
- dplyr  
- plyr  
- cowplot  
- Rtsne  
- FNN
- rARPACK 

### Bioconductor packages

- zinbwave  
- limma  
- edgeR  
- DESeq2  
- BiocParallel  
- doParallel 
- Biobase   
- scde  
- SingleCellExperiment  
- MAST
- [xCell](https://github.com/dviraran/xCell)  
- fgsea  
- clusterExperiment  
- GSEABase  


## Getting started

### Time benchmarking

For each of the real datasets Islam, Usoskin, 10X genomics, respectively run [benchmark_islam.Rmd](https://github.com/statOmics/zinbwaveZinger/blob/master/timebenchmark/islam/benchmark_islam.Rmd), [benchmark_usoskin.Rmd](https://github.com/statOmics/zinbwaveZinger/blob/master/timebenchmark/usoskin/benchmark_usoskin.Rmd), [benchmark_tenx.Rmd](https://github.com/statOmics/zinbwaveZinger/blob/master/timebenchmark/tenx/benchmark_tenx.Rmd). Finally, run [benchmark_all.Rmd](https://github.com/statOmics/zinbwaveZinger/blob/master/timebenchmark/benchmark_all.Rmd).

### Real data

For each of the real datasets analyzed in the paper, there are a `.Rmd` file and a `.R` file in the `real_data` folder, e.g.,
for the Patel data, the files are [patel_covariates.Rmd](https://github.com/drisso/zinb_analysis/blob/master/real_data/patel_covariates.Rmd)
and [patel_plots.R](https://github.com/drisso/zinb_analysis/blob/master/real_data/patel_plots.R).

One needs to compile the .Rmd file first. This will have two effects: (i) it will create an HTML report with useful analyses of
the dataset; and (ii) it will create a `.rda` file with the results of `zinbwave`, `pca`, and `zifa`. Once this file is generated,
one can use the `.R` file to generate the dataset-specific plots found in the paper.

To generate the plots related to silhouette width, one needs to source the [silhouette.R](https://github.com/drisso/zinb_analysis/blob/master/real_data/silhouette.R)
file.

To generate the plots related to the goodness-of-fit, run the `.Rmd` files in the `real_data` folder starting with `goodness_of_fit`, e.g., for the Patel data, the file is [goodness_of_fit_patel.Rmd](https://github.com/drisso/zinb_analysis/blob/master/real_data/goodness_of_fit_patel.Rmd).

### Simulations

To create the simulated datasets from the real datasets used in the paper, first run the code in [simFunction.R](https://github.com/drisso/zinb_analysis/blob/master/sims/figures/simFunction.R). Then, run the `.R` files in the folders in `sims/figures`. Finally, run [figuresPaper.Rmd](https://github.com/drisso/zinb_analysis/blob/master/sims/figures/figuresPaper.Rmd).

To simulate the datasets from the Lun & Marioni model, run [lunSim.R](https://github.com/drisso/zinb_analysis/blob/master/sims/figures/fig6e-g/lunSim.R). It uses file [function.rds](https://github.com/drisso/zinb_analysis/blob/master/sims/figures/fig6e-g/function.rds) generated by the steps described in the Methods section of the paper. Then, run [fitZinbLun.R](https://github.com/drisso/zinb_analysis/blob/master/sims/figures/fig6e-g/fitZinbLun.R).

To fit the simulated datasets with n=10,000 cells, we used a [Makefile](https://github.com/drisso/zinb_analysis/blob/master/sims/figures/fig6ad-S13-S14/Makefile) to launch jobs on a server. Alternatively, you can just call [fitZinb10000.R](https://github.com/drisso/zinb_analysis/blob/master/sims/figures/fig6ad-S13-S14/fitZinb10000.R) from your terminal with the arguments you want.

