---
title: "Usoskin DE analysis"
author: "Koen Van den Berge"
date: "`r Sys.Date()`"
output:
  html_document:
    fig_height: 7
    fig_width: 7
    toc: yes
    code_folding: hide
    toc_float: yes
---

This document uses Fanny's original simulation code in `islam_sims_fc2.Rmd' and has been adapted by Koen to adopt the new simulation framework.

```{r options, echo=FALSE, results="hide",message=FALSE, error=FALSE, include=FALSE, autodep=TRUE, warning=FALSE}
setwd("/Users/koenvandenberge/Dropbox/phdKoen/singleCell/zinbwaveZingerGithub/zinbwaveZinger/realdata/usoskin/")
knitr::opts_chunk$set(fig.align="center", cache=TRUE, error=FALSE, message=FALSE, warning=TRUE)
library(zinbwave)
library(BiocParallel)
library(doParallel)
library(Biobase)
library(edgeR)
library(scales)
library(DESeq2)
library(iCOBRA) # roc
library(limma)
library(genefilter) #filtered pvalues
library(MAST)
library(RColorBrewer)
library(knitr)
load("../../datasets/esetUsoskin.RData")
```


```{r cores}
NCORES <- 2
registerDoParallel(NCORES)
register(DoparParam())
```

```{r design}
cellType= droplevels(pData(eset)[,"Level 3"])
batch = pData(eset)[,"Picking sessions"]
counts = exprs(eset)
keep = rowSums(counts>0)>9
counts=counts[keep,]
```

```{r resultsData}
results = data.frame(edgeR=rep(NA,11),
                    DESeq2=rep(NA,11),
 row.names=c("NF1","NF2","NF3","NF4","NF5","NP1","NP2","NP3","PEP1","PEP2","TH"))
```



# Methods
## RNA-seq methods
### edgeR
```{r edgeR}

  d <- DGEList(counts)
  d <- suppressWarnings(edgeR::calcNormFactors(d))
  design <- model.matrix(~cellType+batch)
  d <- estimateDisp(d, design)
  fit <- glmFit(d,design)
L <- matrix(0,nrow=ncol(fit$coefficients),ncol=11)
rownames(L) <- colnames(fit$coefficients)
colnames(L) <- c("NF1","NF2","NF3","NF4","NF5","NP1","NP2","NP3","PEP1","PEP2","TH")
L[2:11,1] <- -1/10 #NF1 vs. others
L[2:11,2] <- c(1,rep(-1/10,9)) #NF2 vs. others
L[2:11,3] <- c(-1/10,1,rep(-1/10,8)) #NF3 vs. others
L[2:11,4] <- c(rep(-1/10,2),1,rep(-1/10,7)) #NF4 vs. others
L[2:11,5] <- c(rep(-1/10,3),1,rep(-1/10,6)) #NF5 vs. others
L[2:11,6] <- c(rep(-1/10,4),1,rep(-1/10,5)) #NP1 vs. others
L[2:11,7] <- c(rep(-1/10,5),1,rep(-1/10,4)) #NP2 vs. others
L[2:11,8] <- c(rep(-1/10,6),1,rep(-1/10,3)) #NP3 vs. others
L[2:11,9] <- c(rep(-1/10,7),1,rep(-1/10,2)) #PEP1 vs. others
L[2:11,10] <- c(rep(-1/10,8),1,rep(-1/10,1)) #PEP2 vs. others
L[2:11,11] <- c(rep(-1/10,9),1) #TH vs. others
lrtList=list()
for(i in 1:ncol(L)) lrtList[[i]] <- zinbwave::glmWeightedF(fit,contrast=L[,i], ZI=FALSE)
padjListEdgeR=lapply(lrtList, function(x) p.adjust(x$table$PValue,"BH"))
deGenesEdgeR=unlist(lapply(padjListEdgeR,function(x) sum(x<.05)))
deGenesEdgeR
```

### DESeq2
```{r DESeq2}

  colData <- data.frame(cellType = cellType, batch=batch)
  dse <- DESeqDataSetFromMatrix(countData = counts, colData = colData, design = ~ cellType+batch)
  dse <- DESeq2::estimateSizeFactors(dse, type="poscounts")
  dse <- estimateDispersions(dse)
  dse <- nbinomWaldTest(dse, modelMatrixType="standard", betaPrior=TRUE)
resultsNames(dse) #for building contrasts, see ?results
L=matrix(0,nrow=length(resultsNames(dse)),ncol=11)
rownames(L)=resultsNames(dse)
L[2:11,1] <- -1/10 #NF1 vs. others
L[2:11,2] <- c(1,rep(-1/10,9)) #NF2 vs. others
L[2:11,3] <- c(-1/10,1,rep(-1/10,8)) #NF3 vs. others
L[2:11,4] <- c(rep(-1/10,2),1,rep(-1/10,7)) #NF4 vs. others
L[2:11,5] <- c(rep(-1/10,3),1,rep(-1/10,6)) #NF5 vs. others
L[2:11,6] <- c(rep(-1/10,4),1,rep(-1/10,5)) #NP1 vs. others
L[2:11,7] <- c(rep(-1/10,5),1,rep(-1/10,4)) #NP2 vs. others
L[2:11,8] <- c(rep(-1/10,6),1,rep(-1/10,3)) #NP3 vs. others
L[2:11,9] <- c(rep(-1/10,7),1,rep(-1/10,2)) #PEP1 vs. others
L[2:11,10] <- c(rep(-1/10,8),1,rep(-1/10,1)) #PEP2 vs. others
L[2:11,11] <- c(rep(-1/10,9),1) #TH vs. others
resList=list()
for(i in 1:ncol(L)) resList[[i]] = results(dse,contrast=L[,i])
lapply(resList, function(x) sum(x$padj<=0.05, na.rm=TRUE))
```

### limma-voom
```{r limma}
  batchLimma=factor(gsub(batch,pattern="-",replacement="")) #valid names
	design <- model.matrix(~ cellType+batchLimma)
	nf <- suppressWarnings(edgeR::calcNormFactors(counts))
	y <- voom(counts, design, plot = FALSE, lib.size = colSums(counts) * nf)
	fit <- lmFit(y, design)
  contrast.matrix = makeContrasts(
    NF1 = -0.1*cellTypeNF2 - 0.1*cellTypeNF3 - 0.1*cellTypeNF4 - 0.1*cellTypeNF5 - 0.1*cellTypeNP1 - 0.1*cellTypeNP2 - 0.1*cellTypeNP3 - 0.1*cellTypePEP1 - 0.1*cellTypePEP2 - 0.1*cellTypeTH,
    NF2 = cellTypeNF2 - 0.1*cellTypeNF3 - 0.1*cellTypeNF4 - 0.1*cellTypeNF5 - 0.1*cellTypeNP1 - 0.1*cellTypeNP2 - 0.1*cellTypeNP3 - 0.1*cellTypePEP1 - 0.1*cellTypePEP2 - 0.1*cellTypeTH,
    NF3 = -0.1*cellTypeNF2 + cellTypeNF3 - 0.1*cellTypeNF4 - 0.1*cellTypeNF5 - 0.1*cellTypeNP1 - 0.1*cellTypeNP2 - 0.1*cellTypeNP3 - 0.1*cellTypePEP1 - 0.1*cellTypePEP2 - 0.1*cellTypeTH,
    NF4 = -0.1*cellTypeNF2 - 0.1*cellTypeNF3 + cellTypeNF4 - 0.1*cellTypeNF5 - 0.1*cellTypeNP1 - 0.1*cellTypeNP2 - 0.1*cellTypeNP3 - 0.1*cellTypePEP1 - 0.1*cellTypePEP2 - 0.1*cellTypeTH,
    NF5 = -0.1*cellTypeNF2 - 0.1*cellTypeNF3 - 0.1*cellTypeNF4 + cellTypeNF5 - 0.1*cellTypeNP1 - 0.1*cellTypeNP2 - 0.1*cellTypeNP3 - 0.1*cellTypePEP1 - 0.1*cellTypePEP2 - 0.1*cellTypeTH,
    NP1 = -0.1*cellTypeNF2 - 0.1*cellTypeNF3 - 0.1*cellTypeNF4 - 0.1*cellTypeNF5 + cellTypeNP1 - 0.1*cellTypeNP2 - 0.1*cellTypeNP3 - 0.1*cellTypePEP1 - 0.1*cellTypePEP2 - 0.1*cellTypeTH,
    NP2 = -0.1*cellTypeNF2 - 0.1*cellTypeNF3 - 0.1*cellTypeNF4 - 0.1*cellTypeNF5 - 0.1*cellTypeNP1 + cellTypeNP2 - 0.1*cellTypeNP3 - 0.1*cellTypePEP1 - 0.1*cellTypePEP2 - 0.1*cellTypeTH,
    NP3 = -0.1*cellTypeNF2 - 0.1*cellTypeNF3 - 0.1*cellTypeNF4 - 0.1*cellTypeNF5 - 0.1*cellTypeNP1 - 0.1*cellTypeNP2 + cellTypeNP3 - 0.1*cellTypePEP1 - 0.1*cellTypePEP2 - 0.1*cellTypeTH,
    PEP1 = -0.1*cellTypeNF2 - 0.1*cellTypeNF3 - 0.1*cellTypeNF4 - 0.1*cellTypeNF5 - 0.1*cellTypeNP1 - 0.1*cellTypeNP2 - 0.1*cellTypeNP3 + cellTypePEP1 - 0.1*cellTypePEP2 - 0.1*cellTypeTH,
    PEP2 = -0.1*cellTypeNF2 - 0.1*cellTypeNF3 - 0.1*cellTypeNF4 - 0.1*cellTypeNF5 - 0.1*cellTypeNP1 - 0.1*cellTypeNP2 - 0.1*cellTypeNP3 - 0.1*cellTypePEP1 + cellTypePEP2 - 0.1*cellTypeTH,
    TH = -0.1*cellTypeNF2 - 0.1*cellTypeNF3 - 0.1*cellTypeNF4 - 0.1*cellTypeNF5 - 0.1*cellTypeNP1 - 0.1*cellTypeNP2 - 0.1*cellTypeNP3 - 0.1*cellTypePEP1 - 0.1*cellTypePEP2 + cellTypeTH,
  levels=design
)
  fit2 <- contrasts.fit(fit, contrast.matrix)
	fit2 <- eBayes(fit2)
  ttList = list()
  for(i in 1:11) ttList[[i]]  <- topTable(fit2, coef = i, n = nrow(counts), sort.by = "none")
  deGenesLimma = lapply(ttList, function(x) sum(x$adj.P.Val<=0.05))
  deGenesLimma
```

## scRNA-seq methods

### MAST

```{r MAST}
### copied code from FPR_mocks.Rmd on September 14, 2017.
MAST <- function(counts, group, ylim = NULL, xlim = NULL){
  require(MAST)
  tpm <- counts*1e6/colSums(counts)
  tpm <- log2(tpm+1)
  sca <- FromMatrix(tpm,  cData=data.frame(group=group))
  #sca <- FromMatrix(counts,  cData=data.frame(group=group))

  # Adaptive thresholding from MAST vignette
  freq_expressed <- 0.2
  thres <- thresholdSCRNACountMatrix(assay(sca), nbins = 10, min_per_bin = 50, conditions = group)
  #par(mfrow=c(5,4))
  #plot(thres)
  assays(sca) <- list(thresh=thres$counts_threshold, tpm=assay(sca))
  expressed_genes <- freq(sca) > freq_expressed
  sca <- sca[expressed_genes,]

  ngeneson <- apply(counts,2,function(x) mean(x>0))
  CD <- colData(sca)
  CD$ngeneson <- ngeneson
  CD$cngeneson <- CD$ngeneson-mean(ngeneson)
  colData(sca) <- CD
  ## differential expression
  fit <- zlm(~ cngeneson + group , sca = sca)
  lrFit <- lrTest(fit, 'group')
  pval <- lrFit[, 'hurdle', 'Pr(>Chisq)']
  padj <- p.adjust(pval, method = "BH")

  ### MAST filtered the genes, so make a list that is consistent with the original count matrix.
  pvalAll = vector(length=nrow(counts))
  pvalAll[] = 1
  names(pvalAll)=rownames(counts)
  pvalAll[match(names(pval),names(pvalAll))] = pval

  padjAll = vector(length=nrow(counts))
  padjAll[] = 1
  names(padjAll)=rownames(counts)
  padjAll[match(names(padj),names(padjAll))] = padj

  out = cbind(pval = pvalAll, padj = padjAll, logfc = NA)
  return(out)
}

```

### zinbwave_edgeR

```{r}
core <- SummarizedExperiment(counts,
                             colData = data.frame(cellType = cellType, batch=batch))
zinb_c <- zinbFit(core, X = '~ cellType + batch', commondispersion = TRUE, epsilon=1e12)
weights = computeObservationalWeights(zinb_c, counts)
```

### zinbwave-edgeR
```{r zinbwaveedger}

  d <- DGEList(counts)
  d <- suppressWarnings(edgeR::calcNormFactors(d))
  design <- model.matrix(~cellType+batch)
  d$weights = weights
  d <- estimateDisp(d, design)
  fit <- glmFit(d,design)
L <- matrix(0,nrow=ncol(fit$coefficients),ncol=11)
rownames(L) <- colnames(fit$coefficients)
colnames(L) <- c("NF1","NF2","NF3","NF4","NF5","NP1","NP2","NP3","PEP1","PEP2","TH")
L[2:11,1] <- -1/10 #NF1 vs. others
L[2:11,2] <- c(1,rep(-1/10,9)) #NF2 vs. others
L[2:11,3] <- c(-1/10,1,rep(-1/10,8)) #NF3 vs. others
L[2:11,4] <- c(rep(-1/10,2),1,rep(-1/10,7)) #NF4 vs. others
L[2:11,5] <- c(rep(-1/10,3),1,rep(-1/10,6)) #NF5 vs. others
L[2:11,6] <- c(rep(-1/10,4),1,rep(-1/10,5)) #NP1 vs. others
L[2:11,7] <- c(rep(-1/10,5),1,rep(-1/10,4)) #NP2 vs. others
L[2:11,8] <- c(rep(-1/10,6),1,rep(-1/10,3)) #NP3 vs. others
L[2:11,9] <- c(rep(-1/10,7),1,rep(-1/10,2)) #PEP1 vs. others
L[2:11,10] <- c(rep(-1/10,8),1,rep(-1/10,1)) #PEP2 vs. others
L[2:11,11] <- c(rep(-1/10,9),1) #TH vs. others
lrtList=list()
for(i in 1:ncol(L)) lrtList[[i]] <- zinbwave::glmWeightedF(fit,contrast=L[,i])
padjListZinbEdgeR=lapply(lrtList, function(x) p.adjust(x$table$PValue,"BH"))
deGenesZinbEdgeR=unlist(lapply(padjListZinbEdgeR,function(x) sum(x<=.05)))
deGenesZinbEdgeR



```

### zinbwave-DESeq2
```{r zinbwavedeseq2}

  colData <- data.frame(cellType = cellType, batch=batch)
  dse <- DESeqDataSetFromMatrix(countData = counts, colData = colData, design = ~ cellType+batch)
  assays(dse)[["weights"]]=weights
  dse <- DESeq2::estimateSizeFactors(dse, type="poscounts")
  dse <- estimateDispersions(dse)
  dse <- nbinomWaldTest(dse, modelMatrixType="standard", betaPrior=TRUE, useT=TRUE, df=rowSums(weights)-13)
resultsNames(dse) #for building contrasts, see ?results
L=matrix(0,nrow=length(resultsNames(dse)),ncol=11)
rownames(L)=resultsNames(dse)
L[2:11,1] <- -1/10 #NF1 vs. others
L[2:11,2] <- c(1,rep(-1/10,9)) #NF2 vs. others
L[2:11,3] <- c(-1/10,1,rep(-1/10,8)) #NF3 vs. others
L[2:11,4] <- c(rep(-1/10,2),1,rep(-1/10,7)) #NF4 vs. others
L[2:11,5] <- c(rep(-1/10,3),1,rep(-1/10,6)) #NF5 vs. others
L[2:11,6] <- c(rep(-1/10,4),1,rep(-1/10,5)) #NP1 vs. others
L[2:11,7] <- c(rep(-1/10,5),1,rep(-1/10,4)) #NP2 vs. others
L[2:11,8] <- c(rep(-1/10,6),1,rep(-1/10,3)) #NP3 vs. others
L[2:11,9] <- c(rep(-1/10,7),1,rep(-1/10,2)) #PEP1 vs. others
L[2:11,10] <- c(rep(-1/10,8),1,rep(-1/10,1)) #PEP2 vs. others
L[2:11,11] <- c(rep(-1/10,9),1) #TH vs. others
resList=list()
for(i in 1:ncol(L)) resList[[i]] = results(dse,contrast=L[,i])
lapply(resList, function(x) sum(x$padj<=0.05, na.rm=TRUE))

```
